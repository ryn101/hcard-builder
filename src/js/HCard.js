/**
 * Container for HCard specific information
 */
class HCard {
    constructor(
        givenName,
        surname,
        email,
        phone,
        houseNumber,
        street,
        suburb,
        state,
        postcode,
        country,
    ) {
        this.personalDetails = {
            givenName,
            surname,
            email,
            phone,
        };

        this.address = {
            houseNumber,
            street,
            suburb,
            state,
            postcode,
            country,
        };

        Object.defineProperty(this, 'fullName', {
            enumerable: false,
            get: () => [this.personalDetails.givenName, this.personalDetails.surname].filter(Boolean).join(' '),
        });
    }

    /**
     * Returns the full name being a combination of given and surnames
     *
     * @returns {string}
     */
    fullName = () => {
        const { givenName, surname } = this.personalDetails;
        return [givenName, surname].filter(Boolean).join(' ');
    }
}

export default HCard;
