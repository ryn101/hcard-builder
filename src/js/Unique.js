let uniqueMap = {};

/**
 * Class for providing a unique id to classes that extend it
 */
export default class Unique {
    constructor(uniqueIdPrefix) {
        if (typeof uniqueIdPrefix === 'undefined' || uniqueIdPrefix === null) {
            throw new Error('Classes extending Unique must provided a valid id prefix');
        }

        let uniqueIdCounter = uniqueMap[uniqueIdPrefix] || 1;
        this.id = `${uniqueIdPrefix}-${uniqueIdCounter}`;
        uniqueIdCounter += 1;

        uniqueMap[uniqueIdPrefix] = uniqueIdCounter;
    }

    static resetCounters() {
        uniqueMap = {};
    }
}
