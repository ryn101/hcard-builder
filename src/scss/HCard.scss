@import 'colours';
@import 'mixins';

$vcard-padding: 25px;
$photo-wrapper-right: 10px;
$photo-wrapper-width: 83px;

.vcard {
  width: 430px;
  background-color: white;
  border-bottom: 2px solid #9A9A9A;
  height: 306px;
  position: relative;

  .type {
    display: none;
  }

  .banner {
    position: absolute;
    top: -26px;
    right: 5px;
    color: #95A5A6;
  }

  .header {
    height: 100px;
    background-color: $sophisticated-navy;
    padding: 54px $vcard-padding 0 $vcard-padding;

    h1 {
      font-size: 18pt;
      margin: 0;
      color: #fff;
      font-family: Merriweather, Arial, serif;
      width: calc(100% - #{($photo-wrapper-right * 2) + $photo-wrapper-width - $vcard-padding});
      @include ellipsis();
    }
  }

  .photo-wrapper {
    border: 1px solid #9A9A9A;
    border-right-width: 2px;
    border-bottom-width: 2px;
    position: absolute;
    right: $photo-wrapper-right;
    top: 10px;
    height: 104px;
    width: $photo-wrapper-width;
    overflow: hidden;
    text-align: center;
    background-color: #e3e3e3;

    > img {
      height: 100%;
      width: 100%;
      object-fit: cover;
    }
  }

  .details {
    display: block;
    padding: $vcard-padding;
    margin: 0;
    font-style: normal;

    li {
      border-bottom: 1px $on-the-line-grey solid;
      padding-bottom: 1px;
      justify-content: left;

      &:not(:first-child) {
        margin-top: 12px;
      }
    }

    .detail {
      width: 100%;
      overflow: hidden;

      div > span {
        display: inline-block;
      }

      $detail-label-width: 76px;

      .label {
        width: $detail-label-width;

        color: $sophisticated-navy;
        text-transform: uppercase;
        font-size: 7.5pt;
      }

      .value {
        @include ellipsis();
        font-family: Merriweather, Arial, serif;
        width: calc(100% - #{$detail-label-width});
        vertical-align: text-bottom;
        margin-bottom: -5px;
        padding-bottom: 5px;
        min-height: 23px;
      }
    }
  }
}
