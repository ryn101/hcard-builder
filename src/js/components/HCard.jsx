import React from 'react';
import PropTypes from 'prop-types';
import HCardDetails from '../HCardDetails';

import '../../scss/HCard.scss';
import profilePicture from '../../images/profile-default.png';
import HCardProperty from './HCardProperty';
import HCardDetail from './HCardDetail';
import AddressState from '../AddressState';

function HCard(props) {
    const { hCard, avatarFileUrl } = props;
    const streetAddress = [hCard.houseNameOrNumber, hCard.street].filter(Boolean).join(' ');
    const fullName = [hCard.givenName, hCard.surname].filter(Boolean).join(' ');
    const state = new AddressState(hCard.state);

    return (
        <section className="vcard">
            <header className="vcard__header">
                <h2 className="vcard__heading fn" title={fullName}>{fullName}</h2>
                <div className="vcard__photo-wrapper">
                    <img alt="avatar" className="photo" src={avatarFileUrl} />
                </div>
            </header>
            <span className="vcard__banner">HCARD PREVIEW</span>
            <address className="adr">
                <ul className="vcard__details">
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail label="Email" className="flex-row__column">
                            <HCardProperty name="email" value={hCard.email} />
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail label="Phone" className="flex-row__column">
                            <HCardProperty name="tel">
                                <HCardProperty name="type" value="home" />
                                { hCard.phone }
                            </HCardProperty>
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail label="Address" className="flex-row__column">
                            <HCardProperty name="street-address" value={streetAddress} />
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-1">
                        <HCardDetail className="flex-row__column">
                            <HCardProperty name="locality" value={hCard.suburb} />
                            <span>{hCard.suburb && hCard.state ? ', ' : ''}</span>
                            <HCardProperty name="region" value={state.name} abbreviation={state.abbreviation} />
                        </HCardDetail>
                    </li>
                    <li className="vcard__details-row flex-row flex-row--columns-2">
                        <HCardDetail label="Post Code" className="flex-row__column">
                            <HCardProperty name="postal-code" value={hCard.postcode} />
                        </HCardDetail>
                        <HCardDetail label="Country" className="flex-row__column">
                            <HCardProperty name="country" value={hCard.country} />
                        </HCardDetail>
                    </li>
                </ul>
            </address>
        </section>
    );
}

HCard.propTypes = {
    hCard: PropTypes.instanceOf(HCardDetails).isRequired,
    avatarFileUrl: PropTypes.string,
};

HCard.defaultProps = {
    avatarFileUrl: profilePicture,
};

export default HCard;
