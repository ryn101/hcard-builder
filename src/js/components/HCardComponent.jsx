import React, { Component } from 'react';
import PropTypes from 'prop-types';
import HCard from '../HCard';
import HCardDetailRow from '../HCardDetailRow';
import HCardDetailRowComponent from './HCardDetailRowComponent';
import HCardProperty, { HCardPropertyType } from '../HCardProperty';

import '../../scss/HCard.scss';
import profilePicture from '../../images/profile-default.png';

class HCardComponent extends Component {
    constructor() {
        super();

        this.hCardProps = Object.values(HCardPropertyType).reduce((result, name) => {
            // eslint-disable-next-line no-param-reassign
            result[name] = new HCardProperty(name);

            return result;
        }, {});

        this.hCardRows = [
            new HCardDetailRow().addDetail('Email', this.hCardProps[HCardPropertyType.email]),
            new HCardDetailRow().addDetail('Phone', this.hCardProps[HCardPropertyType.telephone]),
            new HCardDetailRow().addDetail('Address', this.hCardProps[HCardPropertyType.streetAddress]),
            new HCardDetailRow().addDetail('', [
                this.hCardProps[HCardPropertyType.locality],
                this.hCardProps[HCardPropertyType.region],
            ], ', '),
            new HCardDetailRow()
                .addDetail('Postcode', this.hCardProps[HCardPropertyType.postalCode])
                .addDetail('Country', this.hCardProps[HCardPropertyType.country]),
        ];
    }

    /**
     * Updates the hCardProperty values
     *
     * @param hCard
     * @returns HCardDetailRow[]
     */
    updateHCardProperties = (hCard) => {
        const { email, phone } = hCard.personalDetails;
        const {
            houseNumber,
            street,
            suburb,
            state,
            postcode,
            country,
        } = hCard.address;

        // noinspection JSPrimitiveTypeWrapperUsage
        this.hCardProps[HCardPropertyType.email].value = email;
        this.hCardProps[HCardPropertyType.telephone].value = phone;
        this.hCardProps[HCardPropertyType.streetAddress].value = [houseNumber, street].filter(Boolean).join(' ');
        this.hCardProps[HCardPropertyType.locality].value = suburb;
        this.hCardProps[HCardPropertyType.region].value = state;
        this.hCardProps[HCardPropertyType.postalCode].value = postcode;
        this.hCardProps[HCardPropertyType.country].value = country;
    };

    /**
     * Render the individual detail row components
     *
     * @param detailRows
     * @returns {HCardDetailRowComponent}
     */
    renderDetails = (detailRows) => detailRows.map((detailRow) => (
        <HCardDetailRowComponent key={detailRow.id} hCardDetailRow={detailRow} />
    ));

    render() {
        const { hCard, avatarFileUrl } = this.props;
        this.updateHCardProperties(hCard);

        return (
            <div className="vcard">
                <span className="banner">HCARD PREVIEW</span>
                <div className="header">
                    <h1 className="fn" title={hCard.fullName}>{ hCard.fullName }</h1>
                    <div className="photo-wrapper">
                        <img alt="avatar" className="photo" src={avatarFileUrl} />
                    </div>
                </div>
                <div className="adr details">
                    <ul>
                        {
                            this.renderDetails(this.hCardRows)
                        }
                    </ul>
                </div>
            </div>
        );
    }
}

HCardComponent.propTypes = {
    hCard: PropTypes.instanceOf(HCard).isRequired,
    avatarFileUrl: PropTypes.string,
};

HCardComponent.defaultProps = {
    avatarFileUrl: profilePicture,
};

export default HCardComponent;
