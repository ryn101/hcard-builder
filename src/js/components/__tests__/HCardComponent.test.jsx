import React from 'react';
import ReactDOM from 'react-dom';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import HCard from '../../HCard';
import HCardComponent from '../HCardComponent';
import Unique from '../../Unique';

configure({ adapter: new Adapter() });

beforeAll(() => {
    Unique.resetCounters();
});

it('renders the hCard without crashing', () => {
    const hCard = new HCard();
    const div = document.createElement('div');
    ReactDOM.render(<HCardComponent hCard={hCard} />, div);
    ReactDOM.unmountComponentAtNode(div);
});

describe('once the hcard has rendered', () => {
    let hCardNode;
    let hCard;

    beforeEach(() => {
        hCard = new HCard();
        hCardNode = shallow(<HCardComponent hCard={hCard} />);
    });

    it('renders with state correctly', () => {
        expect(hCardNode).toMatchSnapshot();
    });

    it('renders the full name correctly', () => {
        const givenName = 'Ryan';
        const surname = 'O\'Connor';

        hCard.personalDetails.givenName = givenName;
        hCard.personalDetails.surname = surname;
        hCardNode.setProps({ hCard });
        hCardNode.update();

        const fullNameNode = hCardNode.find('.fn');
        expect(fullNameNode.text()).toEqual(`${givenName} ${surname}`);
    });

    it('set the avatar file url', () => {
        const avatarFileUrl = '/images/test.png';
        hCardNode.setProps({ avatarFileUrl });
        hCardNode.update();

        const photoNode = hCardNode.find('.photo');
        expect(photoNode.prop('src')).toEqual(avatarFileUrl);
    });
});
