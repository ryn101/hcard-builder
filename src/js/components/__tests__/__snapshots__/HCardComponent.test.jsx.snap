// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`once the hcard has rendered renders with state correctly 1`] = `
<div
  className="vcard"
>
  <span
    className="banner"
  >
    HCARD PREVIEW
  </span>
  <div
    className="header"
  >
    <h1
      className="fn"
      title=""
    />
    <div
      className="photo-wrapper"
    >
      <img
        alt="avatar"
        className="photo"
        src="test-file-stub"
      />
    </div>
  </div>
  <ul
    className="adr details"
  >
    <HCardDetailRowComponent
      hCardDetailRow={
        HCardDetailRow {
          "details": Array [
            HCardDetail {
              "id": "hcd-7",
              "label": "Email",
              "properties": Array [
                HCardProperty {
                  "name": "email",
                  "value": undefined,
                },
              ],
              "propertiesSeparator": undefined,
            },
          ],
          "id": "hcdr-6",
        }
      }
      key="hcdr-6"
    />
    <HCardDetailRowComponent
      hCardDetailRow={
        HCardDetailRow {
          "details": Array [
            HCardDetail {
              "id": "hcd-8",
              "label": "Phone",
              "properties": Array [
                HCardProperty {
                  "name": "tel",
                  "value": undefined,
                },
              ],
              "propertiesSeparator": undefined,
            },
          ],
          "id": "hcdr-7",
        }
      }
      key="hcdr-7"
    />
    <HCardDetailRowComponent
      hCardDetailRow={
        HCardDetailRow {
          "details": Array [
            HCardDetail {
              "id": "hcd-9",
              "label": "Address",
              "properties": Array [
                HCardProperty {
                  "name": "street-address",
                  "value": "",
                },
              ],
              "propertiesSeparator": undefined,
            },
          ],
          "id": "hcdr-8",
        }
      }
      key="hcdr-8"
    />
    <HCardDetailRowComponent
      hCardDetailRow={
        HCardDetailRow {
          "details": Array [
            HCardDetail {
              "id": "hcd-10",
              "label": "",
              "properties": Array [
                HCardProperty {
                  "name": "locality",
                  "value": undefined,
                },
                HCardProperty {
                  "name": "region",
                  "value": undefined,
                },
              ],
              "propertiesSeparator": ", ",
            },
          ],
          "id": "hcdr-9",
        }
      }
      key="hcdr-9"
    />
    <HCardDetailRowComponent
      hCardDetailRow={
        HCardDetailRow {
          "details": Array [
            HCardDetail {
              "id": "hcd-11",
              "label": "Postcode",
              "properties": Array [
                HCardProperty {
                  "name": "postal-code",
                  "value": undefined,
                },
              ],
              "propertiesSeparator": undefined,
            },
            HCardDetail {
              "id": "hcd-12",
              "label": "Country",
              "properties": Array [
                HCardProperty {
                  "name": "country",
                  "value": undefined,
                },
              ],
              "propertiesSeparator": undefined,
            },
          ],
          "id": "hcdr-10",
        }
      }
      key="hcdr-10"
    />
  </ul>
</div>
`;
